### PR TITLE
Add MinIO Rock-on #231

### PR DESCRIPTION
**Pull request by MikeMcPherson**
> To ease and accelerate the PR submission, please use the template below to provide all requested information.
> You can find guidelines on which docker image to use in the [Rockstor's documentation](http://rockstor.com/docs/contribute_rockons.html#which-docker-image-should-i-use), 
> as well as examples of proper formatting in other rock-ons ([here](https://github.com/rockstor/rockon-registry/blob/master/teamspeak3.json)
> and [here](https://github.com/rockstor/rockon-registry/blob/master/nextcloud-official.json))

Fixes # 231.

### General information on project
This pull request proposes to add a new rock-on for the following project:
- name: MinIO
- website: https://min.io/
- description: MinIO is an S3-compatible object storage server.  Requires Rockstor 3.9.2-39 or later.

### Information on docker image
- docker image: https://hub.docker.com/r/minio/minio
- is an official docker image available for this project?: yes

### Checklist
- [X] Passes [JSONlint](https://jsonlint.com) validation
- [X] Entry added to `root.json` in _alphabetical_ order (for new rock-on only)
- [X] `"description"` object lists and links to the docker image used
- [X] `"description"` object provides information on the image's particularities (advantage over another existing rock-on for the same project, for instance)
- [X] `"website"` object links to project's main website

### CODE REVIEW

**Comment by FroggyFlox** [comment]
Hi @MikeMcPherson , 

As mentioned in my review, everything seems to behave as intended with the caveat that I couldn't test whether a MinIO client could successfully connect to this server or not.

The only thing I would like to see is the mention and link to the docker image used (see my code suggestion), otherwise I believe it is ready to be merged. Immense thanks for writing a documentation for it as well, that is excellent! I'll get to that PR once we have this rock-on merged and published.

Thanks again!

**Comment by MikeMcPherson** [comment]
I think I accepted your addition and did a commit.  Let me know if it didn't work.

**Comment by FroggyFlox** [comment]
> 
> 
> I think I accepted your addition and did a commit. Let me know if it didn't work.

Thanks, I do see it! I'll give it a final test and then I believe it should be good to go after that.

**Comment by FroggyFlox** [comment]
Hi @MikeMcPherson ,

First of all, please accept my apologies for not providing any update on this; I've been working with @phillxnet on allowing me to better test and work in this repository and I'm sorted enough now that I could test your PR properly. Indeed, while I could confirm that all was working fine on x86_64 machines, I still needed to test and verify the aarch64 side of things.. which I now just could.

Unfortunately, things seem to be a little more complicated on the aarch64 platform, but nothing related to your PR itself. The problem lies with the fact that the MinIO official images do not appear to be multi-arch when pulling the `latest` tag. They do release dedicated tags for aarch64 but only when using a **release-specific tag**. For instance, a tag compatible with the Raspberry Pi would be: `RELEASE.2020-10-12T21-53-21Z-arm64` with the x86_64 counterpart `RELEASE.2020-10-12T21-53-21Z`.

This means this image cannot currently be used as a true multi-arch image for the moment. Although their community has created a multi-arch image (https://hub.docker.com/r/jessestuart/minio/), we do prefer sticking with official images like you did when we can so I would lean towards rephrasing this rock-on description to better reflect the compatibility with x86_64/amd64 machines only. @phillxnet , what do you think?

Of note, it seems the MinIO team is not in a position to create such multi-arch image so an official one will probably not come any time soon:
https://github.com/minio/minio/issues/9546#issuecomment-710714762

We are planning on improving how we deal with such disparities in architecture for our rock-ons in the future, so we might be able to revisit this later.

@MikeMcPherson , @phillxnet , what do you think?

**Comment by phillxnet** [comment]
@FroggyFlox Re:
> I would lean towards rephrasing this rock-on description to better reflect the compatibility with x86_64/amd64 machines only. @phillxnet , what do you think?

Agreed. It's a simple work around until we can add custom tags or surface arch compatibility within the Rock-on interface, and helps us not go backwards on our commitments re: 
https://github.com/rockstor/rockstor-core/issues/2191

So am I wright in thinking we just want to include in the description, say "(x86_64 only)", and then we can pop this in along with it's doc. @FroggyFlox Do I have that correct.

**Comment by FroggyFlox** [comment]
> So am I wright in thinking we just want to include in the description, say "(x86_64 only)", and then we can pop this in along with it's doc. @FroggyFlox Do I have that correct.

That's what I was thinking, yes.

**Comment by MikeMcPherson** [comment]
This works for me, as well.  Feel free to make the change as you see fit.

 

I’m sorry I didn’t test on an aarch64 machine.  Honestly, I didn’t think of it.

 

Mike

File: minio.json
Comment (FroggyFlox) [inline minio.json:3]: Detail architecture supported:
```suggestion
        "description": "MinIO is an S3-compatible object storage server. This rock-on requires Rockstor 3.9.2-39 or later and is compatible with x86_64/amd64 machines only.<p>Based on the official MinIO docker image: <a href='https://hub.docker.com/r/minio/minio' target='_blank'>https://hub.docker.com/r/minio/minio</a>.",
```

**Review by FroggyFlox — APPROVED** [review]
@MikeMcPherson , 

Thanks a lot for getting back so quickly despite my very long delay in doing so myself.

Following your approval, I went ahead and made the small changes to the description. I'll try to get this in as quickly as I can.

Thanks again for your continued effort on this one, that is a really good addition to the project.

**Comment by MikeMcPherson** [comment]
Thanks to all of you!  You do the heavy lifting.

 

Mike

**Comment by FroggyFlox** [comment]
@MikeMcPherson ,

I'm going to merge your PR now, as all still looks good on my end. Thanks a lot for your sustained efforts in contributing to the project! I'll take care of your related PR in our documentation repository (https://github.com/rockstor/rockstor-doc/pull/245#issue-456734862) soon after.

Thanks!

**Comment by FroggyFlox** [comment]
@MikeMcPherson , 

I forgot to mention that I will take care of publishing this change as soon as I can (hopefully tomorrow); it will then be visible in the list of available Rock-ons for all Rockstor users.
I will make an announcement on the forum when that is taken care of.

Thanks again,

**Comment by harshavardhana** [comment]
> Of note, it seems the MinIO team is not in a position to create such multi-arch image so an official one will probably not come any time soon:
> [minio/minio#9546 (comment)](https://github.com/minio/minio/issues/9546#issuecomment-710714762)

The latest release now has support for multiarch manifests

**Comment by FroggyFlox** [comment]
That is great news, @harshavardhana ! I'm glad to see myself proven wrong!

Thanks a lot for keeping us updated, we really appreciate it!

**Pull request merged**
